src/internal/connector/onedrive/url_cache.go

```diff
@@ -0,0 +1,348 @@
+	}
+
+	if refreshInterval < 0 {
+		return clues.New("invalid refresh interval")
+	}
+
+	if driveEnumerator == nil {
+		return clues.New("nil drive enumerator")
+	}
+
+	if svc == nil {
+		return clues.New("nil graph servicer")
+	}
+
+	if itemPagerFunc == nil {
+		return clues.New("nil item pager")
+	}
+
+	return nil
+}
+
+// getItemProps returns the item properties for the specified drive item ID
+func (uc *urlCache) getItemProperties(
+	ctx context.Context,
+	itemID string,
+) (*itemProperties, error) {
+	if len(itemID) == 0 {
+		return nil, clues.New("item id is empty")
+	}
+
+	ctx = clues.Add(ctx, "drive_id", uc.driveID)
+
+	// Lazy refresh
+	if uc.needsRefresh() {
+		err := uc.refreshCache(ctx)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	url, err := uc.readCache(ctx, itemID)
+	if err != nil {
+		return nil, err
+	}
+
+	return url, nil
+}
+
+// needsRefresh returns true if the cache is empty or if refresh interval has
+// elapsed
+func (uc *urlCache) needsRefresh() bool {
+	uc.cacheLock.RLock()
+	defer uc.cacheLock.RUnlock()
+
+	return len(uc.urlMap) == 0 || time.Since(uc.lastRefreshTime) > uc.refreshInterval
+}
+
+// refreshCache refreshes the URL cache by performing a delta query.
+func (uc *urlCache) refreshCache(
+	ctx context.Context,
+) error {
+	// Acquire mutex to prevent multiple threads from refreshing the
+	// cache at the same time
+	uc.refreshMutex.Lock()
+	defer uc.refreshMutex.Unlock()
+
+	// If the cache was refreshed by another thread while we were waiting
+	// to acquire mutex, return
+	if !uc.needsRefresh() {
+		return nil
+	}
+
+	// Hold cache lock in write mode for the entire duration of the refresh.
+	// This is to prevent other threads from reading the cache while it is
+	// being updated page by page
+	uc.cacheLock.Lock()
+	defer uc.cacheLock.Unlock()
+
+	// Issue a delta query to graph
+	logger.Ctx(ctx).Debugw("refreshing url cache")
+
+	err := uc.deltaQuery(ctx)
+	if err != nil {
+		return err
+	}
+
+	logger.Ctx(ctx).Debugw("url cache refresh complete")
+	// Update last refresh time
+	uc.lastRefreshTime = time.Now()
+
+	return nil
+}
+
+// deltaQuery will perform a delta query on the drive and update the cache
+// TODO: Check if this function is adding any value?
+// Remove it and use collectDriveItems directly
+func (uc *urlCache) deltaQuery(
+	ctx context.Context,
+) error {
+	driveEnumerator := uc.driveEnumerator
+	if driveEnumerator == nil {
+		driveEnumerator = collectDriveItems
+	}
+
+	logger.Ctx(ctx).Debugw("Starting delta query")
+
+	err := driveEnumerator(
+		ctx,
+		uc.itemPagerFunc(uc.svc, uc.driveID, ""),
+		uc.updateCache,
+		"",
+		uc.Errors,
+	)
+	if err != nil {
+		return clues.Wrap(err, "delta query failed").WithClues(ctx)
+	}
+
+	uc.deltaQueryCount++
+
+	return nil
+}
+
+// collectDriveItems will enumerate all items in the specified drive and hand
+// them to the provided `collector` method
+// TODO: This is a clone of collectItems call. Refactor collectItems to remove
```

Comment:
What needs to be changed in `collectItems`?

Reply:
oops! just saw this. I actually refactored the cache code to rely on existing `collectItems` as we had discussed. 
this felt like a better approach than refactoring `collectItems`.